Web publish plugin

seinecle opened the pull request:
## New plugin or plugin update?

* [X] New Plugin
* [ ] Update

## What is the purpose of this plugin?
Enable the user to publish the network they are working on, to the web. See this issue for details of the implementation:
https://github.com/gephi/gephi-plugins/issues/262#issuecomment-1231627948
  
## How to test your plugin in Gephi?
1. Go to File -> Export -> Publish to the Web
2. Follow the instructions in the "Setup" tab
3. Switch to the "Publish" tab and publish.

## Checklist before submission

* [X] Did you merged with the `master` branch to get the latest updates?
* [X] Did you build and test the plugin successfully locally?
* [X] Did you include metadata (e.g. plugin author, license) in the `pom.xml` file
* [X] Did you make sure NOT to include any unecessary files in your PR?
* [ ] Did you write unit tests to test your plugin functionalities?

mbastian commented:
Nice @seinecle and sorry for the delay. I started looking into it and have some general comments and questions:

1. You currently don't configure the GEXF exporter. Are all the default settings what you would expect? For instance I don't think the viewer support dynamic graphs so we should probably turn this off.
2. How do you deal with [Gist's limits](https://docs.github.com/en/rest/gists/gists#truncation)? Seems like the limit could be reach fairly quickly with large graphs so we should probably deal with it. At least inform the user. Maybe we should also thing about strategies to reduce the file size? Maybe Sigma.js supports zipped's GEXF?
3. You have a number of hardcoded strings in your code like URLs, clientId etc. Could you turn all of those into constants for code clarity? (private static final String)

seinecle commented:
hi  @mbastian,

1. I didn't know the exporter can be configured. I feel like the export to gist should preserve the network exactly as the user is working on it in Gephi, so I would not restrain / limit the gexf to some features. That's rather up to Retina / Sigma to handle it in the way they need?
2. I was not aware of these limits. I'll have a look and yes, I will add warnings and will add ways to handle the exceptions related to limits
3. Good point, will do.

seinecle commented:
Did some verifications:

Gist has a [push limit at 100Mb per file](https://docs.github.com/en/repositories/working-with-files/managing-large-files/about-large-files-on-github) (when not using GLFS).

I used the plugin to publish a network which corresponds to a gexf file of 61Mb with 2000 nodes and 279,000 edges.

- the upload on Gist works fine, as expected
- Retina can't display it: no error msg, does the page doesn't load

During the Gephi week I had asked @jacomyal about rough upper limits for the size of a network to be displayed by Retina. It would be around 10k nodes and 20k edges - above this, the browser might get slower on some computers.

So it seems that the first limit is the network size that Retina can handle, not the max file size Gist accepts.

I propose that we set a limit to the size of the network to be published? Like: 20k nodes max, 40k edges max? If any of these limits is reached, the "Publish" button remains greyed out with an explainer.

Your comments are welcome @mbastian @jacomyal

mbastian commented:
@seinecle Thanks for verifying! It's a good news although it does contradict the Gist documentation, which I find strange. The link you posted is for GitHub and may not apply to Gist. 

Regarding the size limit I'm not a fan of forbidding the upload but maybe rather inform the user that the experience is not gonna be optimal. The node/edge counts is also only partially a factor in the file size. The number of attributes and their content can also be a bit factor.

seinecle commented:
Ok, so no hard limit but a warning then - I'll update the plugin today.

mbastian commented on modules/WebPublishPlugin/src/main/java/net/clementlevallois/web/publish/plugin/controller/GephiPluginDesktopLogic.java:
```diff
@@ -0,0 +1,49 @@
+ * @author Clement Levallois
+ */
+@ServiceProvider(service = ExporterClassUI.class)
+public final class GephiPluginDesktopLogic implements ExporterClassUI {
```
Poor class name. Something more straightforward like `WebPublishExporterUI` would be clearer.

seinecle replied:
✅

mbastian commented on modules/WebPublishPlugin/src/main/java/net/clementlevallois/web/publish/plugin/controller/GephiPluginDesktopLogic.java:
```diff
@@ -0,0 +1,49 @@
+
+ * @author Clement Levallois
+ */
+@ServiceProvider(service = ExporterClassUI.class)
```
You may want to add a position to make sure it always appear third in the menu, after the default options. position=200 or something like this.

seinecle replied:
✅

mbastian commented on modules/WebPublishPlugin/src/main/resources/net/clementlevallois/web/publish/plugin/controller/Bundle.properties:
```diff
@@ -0,0 +1,47 @@
+general.message.error.empty_network=Your network is empty. Create a network to publish it.
+JPanelWebExport.step2.title=2. Login to your Github Account
+JPanelWebExport.step1.title=1. You need a Github account and be logged in
+general.message.plugin.name=Publish to the web
```
As a convention, all actions menu that open a new menu/window should end with three dots "..."

seinecle replied:
✅

mbastian commented on modules/WebPublishPlugin/pom.xml:
```diff
@@ -0,0 +1,127 @@
+                <groupId>org.apache.netbeans.utilities</groupId>
+                <artifactId>nbm-maven-plugin</artifactId>
+                <configuration>
+                    <licenseName>Apache 2.0</licenseName>
```
licenseFile is missing

seinecle replied:
✅

mbastian commented on modules/WebPublishPlugin/src/main/java/net/clementlevallois/web/publish/plugin/controller/GlobalConfigParams.java:
```diff
@@ -0,0 +1,23 @@
+/*
```
Delete those comments

mbastian commented on modules/WebPublishPlugin/src/main/java/net/clementlevallois/web/publish/plugin/controller/JPanelWebExport.java:
```diff
@@ -0,0 +1,612 @@
+ * https://github.com/gephi/gephi-plugins/issues/262#issuecomment-1231627948
+ *
+ */
+public class JPanelWebExport extends javax.swing.JPanel {
```
I wouldn't start a class name with JPanel. Rather start with the class purpose and finish by panel or JPanel.

mbastian commented on modules/WebPublishPlugin/src/main/java/net/clementlevallois/web/publish/plugin/controller/JPanelWebExport.java:
```diff
@@ -0,0 +1,612 @@
+
+    private static final ResourceBundle bundle = NbBundle.getBundle(GephiPluginDesktopLogic.class);
+
+    public static final String COLOR_SUCCESS = "#45ba48";
```
You have quite a bit of colors hardcoded here, and then Color.WHITE. Not a big deal but with dark mode this may cause issues

mbastian commented on modules/WebPublishPlugin/src/main/java/net/clementlevallois/web/publish/plugin/controller/JPanelWebExport.java:
```diff
@@ -0,0 +1,612 @@
+
+                boolean success = false;
+                long startTime = System.currentTimeMillis();
+                long maxDuration = 900_000;
```
Is that 15 minutes? that sounds like a really large timeout

mbastian commented on modules/WebPublishPlugin/src/main/java/net/clementlevallois/web/publish/plugin/controller/JPanelWebExport.java:
```diff
@@ -0,0 +1,612 @@
+                }
+            } catch (IOException | InterruptedException ex) {
+                Exceptions.printStackTrace(ex);
+                jTextFieldGithubErrorMsg.setText(ex.getMessage());
```
Within the `SwingWorker` class you're running on a separate thread than the EDT (the swing thread). All the Swing code should normally be sent back to the EDT, usually with `SwingUtilities.invokeLater()`. It will work but not fully correct.

mbastian commented on modules/WebPublishPlugin/src/main/java/net/clementlevallois/web/publish/plugin/controller/JPanelWebExport.java:
```diff
@@ -0,0 +1,612 @@
+        protected Void doInBackground() throws Exception {
+            JsonObject jsonObject = new JsonObject();
+            try {
+                HttpClient client = HttpClient.newHttpClient();
```
Did you try what this does when one is not connected to the internet?

mbastian commented on modules/WebPublishPlugin/src/main/java/net/clementlevallois/web/publish/plugin/controller/PublishingActions.java:
```diff
@@ -0,0 +1,147 @@
+
+    private static final int BYTES_IN_A_MEGABYTE = 1_048_576;
+
+    public static JsonObject getGexfAsString() {
```
I don't get why you're using a `JsonObject` to pass data between this class and the client. Seems like your goal was to be able to pass the gexf data plus some kind of error codes with it. 
- The GEXF string seems to be converted back and forth multi times between JSON and String as a result
- Instead of error codes, a more standard way in Java would be to create your own exception

mbastian commented on modules/WebPublishPlugin/src/main/java/net/clementlevallois/web/publish/plugin/controller/PublishingActions.java:
```diff
@@ -0,0 +1,147 @@
+            return jsonObject;
+        }
+
+        if (graph.getNodeCount() > 10_000 || graph.getEdgeCount() > 20_000) {
```
Those limits are hardcoded, not ideal IMO

mbastian commented on modules/WebPublishPlugin/src/main/java/net/clementlevallois/web/publish/plugin/controller/PublishingActions.java:
```diff
@@ -0,0 +1,147 @@
+
+        try {
+            HttpClient client = HttpClient.newHttpClient();
+            String url = "https://api.github.com/gists";
```
Hardcoded URL

seinecle replied:
If the URL ever changes, the query parameters might change as well. For this reason, I would rather keep in the same view / location both the URL and the query params.

mbastian reviewed (changes requested):
I added some comments, mostly related to conventions. One general comment about the architecture is trying to better decouple the business logic from the UI. You have part of it in the `PublishingActions`, which I think is great as it allows testability. You can check the 3 unit tests I created for reference. On the other hand, some other logic is buried somewhere in the `JPanelWebExport` class, making it difficult to unit test. Ideally all logic is wrapped into clear functions that can be unit tested.

seinecle commented:
Wow indeed many improvements to make. My comments:

- on the json object that I use to pass stuff around. It is indeed weird and not ideal. You suggest that I could throw exceptions instead. Ok, good! I have never worked this way though, so I should get back to some Java core textbook. Will take a couple of days.
- on the SwingWorker / EDT stuff: will go and try what you suggest. I am not familiar at all with these things so any code example / resource would be great, by any chance.

mbastian commented:
@seinecle I did a deeper refactoring to help decouple the UI from the business logic, which was still tightly coupled. In addition, I uses our LongTask API instead of SwingWorker. With that, you easily get the progress bar too. The two `Runnable` classes are now completely decoupled from the UI. The UI updates itself through a `LongTaskListener` (success) and `LongTaskErrorHandler` (error).

One UI improvement I think you could still make is on the final result. It's really paintful to have to copy-paste the final URL from a text field. Why not having an "Open" button or "Copy to clipboard"? Java can easily open the browser for you with an URL.

seinecle commented:
@mbastian thanks! `LongTask `seems super nice, glad to discover this utility! And thanks for the refactoring.

I have:

- added the buttons you mention (open in web browser + copy link to clipboard).
- this needed to change slightly the way the `LongTask` was returning strings
- localized all strings
- renamed a class to WebExport JPanel

mbastian commented:
Thanks @seinecle! I just noticed that the access token we get from GitHub expires after 8 hours. This isn't something we can change but there is a [flow we can use via refresh tokens](https://docs.github.com/en/developers/apps/building-github-apps/refreshing-user-to-server-access-tokens) to have 6 months tokens. It seems unpractical to have to go through the authentification every day so it would be good to implement that. We shall also change the narrative in the UI. It's a little more complicated than "Setup once".

seinecle commented:
ok, thanks for identifying this! Will have a look.

seinecle commented:
@mbastian reading the doc, it seems we can just opt out of expiring tokens? If I understand correctly, the 8 hours limit is just for expiring tokens.
So in Gephi's github app ("gephi-lite"), there just needs to opt out expiring tokens. The steps to find this option are detailed in the link you included in your comment. Please let me know if I misunderstand something.

mbastian commented:
@seinecle Interesting. It still don't really get from the doc what happens when we opt-out? Access token don't expire at all? I opted out now. Let's see how it goes

seinecle commented:
@mbastian do you think we can move to the publication?

mbastian commented:
@seinecle Yes, almost :) Looks like you inverted the buttons between gist view versus network viz.

mbastian commented:
@seinecle The plugin is published https://gephi.org/plugins/#/plugin/web-publish-plugin

Thanks for this great work! cc @jacomyal

seinecle commented:
@mbastian  Thank you for the extensive code review !!

DeJeroenBakker commented:
First of all, thanks for developing this plugin @seinecle. I can imagine this being very useful! 

The plugin doesn't seem to work for me at the moment, however. The authentication process works flawlessly, but when I try to publish my network, I always get the following 500 error:

`Error creating online gexf file. Error code:500; Error message:` 

Is there any way to fix this? I've tried re-authenticating, using a different network and restarting Gephi, but it doesn't seem to make a difference.

Thanks!
